### DISCUSSION TITLE
Fixed version lock on pydantic-core

### DISCUSSION THREAD

**Discussion by eskultety** [pydantic/pydantic#10670] [Question] [answered]
Hey team,
We use dependabot for tracking and updating dependencies in our project and I've been meaning to ask this for a long time since the dependabot update constantly chokes on `pydantic`'s `pydantic-core` version requirements:
```
INFO: pip is looking at multiple versions of pydantic to determine which version is compatible with other requirements. This could take a while.

The conflict is caused by:
    The user requested pydantic-core==2.24.2
    pydantic 2.9.2 depends on pydantic-core==2.23.4
```
This led us to even split pydantic dependency update to a separate dependabot group so that other dependencies in the same PR would not be blocked on the CI failure.

What's the reason for locking your dependency on `pydantic-core` to a particular version instead of using `>=`  an '<' to allow some breathing room when it comes to minor and micro releases which (provided you follow SemVer) must not incorporate any breaking changes.

https://github.com/pydantic/pydantic/blob/a1ffc82f943f856c8f3ba2f1e709b00cbdadc11d/pyproject.toml#L54

**Comment by Viicos** [comment, marked as answer]
Each `pydantic-core` version is meant to work with a single `pydantic` version (we don't follow SemVer for `pydantic-core`).

I'll note that iirc dependabot is not smart enough to understand constraints between packages. Say you have `p1` (with the latest version being `3.2.1`) depending on `p2<=2.0.0` (the latest version being `1.10.0`). In your lock file, you'll end up with `p1==3.2.1` and `p2==1.10.0`. `p2` version `2.0.0` gets released at some point. Dependabot will happily upgrade `p2` to `2.0.0` without knowing the constraint from `p1`. 

Ideally, `pydantic-core` should just be bundled with `pydantic`, but this is a bigger task.

**Reply by eskultety** [reply]
> depending on p2<=2.0.0 (the latest version being 1.10.0). In your lock file, you'll end up with p1==3.2.1 and p2==1.10.0. p2 version 2.0.0 gets released at some point. Dependabot will happily upgrade p2 to 2.0.0 without knowing the constraint from p1.

But in your example *2.0.0* still satisfies the `<=2.0.0` constraint, so this is still okay to do from dependabot's POV. I can see your point though since this is still apparently the case (https://github.com/dependabot/dependabot-core/issues/5140) and so I'll probably just put `pydantic-core` to the list of dep excludes for dependabot given that:

> Each pydantic-core version is meant to work with a single pydantic version

**Reply by Viicos** [reply]
Sorry, I meant something like `p2` version `2.0.1` gets released.
